File: Himark/core/order.py
"""The spelling order: shortlex over spellings, and symbolic interval sets.

FOUNDATION.md fixes the **spelling order** as shortlex (shorter first, ties by
code point) -- a well-order of type omega. This module realizes it three ways:

- :func:`spelling_index` places every spelling at its global shortlex position,
  a big integer; :func:`spelling_from_index` inverts it and :func:`successor`
  steps it by one. Together they make the order arithmetic rather than iterative.
- :class:`IntervalSet` is a normalized set of half-open shortlex intervals over
  those positions. A final segment or range denotes as an ``IntervalSet``; even
  an astronomically large finite range is a single interval, never materialized.

Every operation is derived from the integer positions, so a set that is infinite
(an unbounded final segment) still answers ``contains``/``rank`` in finite time.
"""

from __future__ import annotations

from dataclasses import dataclass

# Base of the shortlex enumeration: the whole Unicode code space, 0..0x10FFFF.
# Surrogates ride along by design (FOUNDATION.md); nothing is special-cased.
_N = 0x110000
_MAX = 0x10FFFF


def spelling_key(s: str) -> tuple[int, str]:
    """Shortlex comparison key: shorter first, ties broken code point by code point."""
    return (len(s), s)


def successor(s: str) -> str:
    r"""Return the shortlex successor: the unique next spelling after ``s``.

    Increments the last non-maximal code point and zeroes everything after it;
    an all-maximal string rolls to ``"\x00" * (len(s) + 1)``. Total: every
    spelling has a successor, so ``spelling_index(successor(s)) == index + 1``.
    """
    codes = [ord(c) for c in s]
    for i in range(len(codes) - 1, -1, -1):
        if codes[i] != _MAX:
            codes[i] += 1
            for j in range(i + 1, len(codes)):
                codes[j] = 0
            return "".join(chr(c) for c in codes)
    return "\x00" * (len(s) + 1)


def _shorter_count(length: int) -> int:
    """Number of spellings shorter than ``length``: the geometric sum sum_{j<len} N^j."""
    return (_N**length - 1) // (_N - 1) if length else 0


def spelling_index(s: str) -> int:
    """Return the global shortlex position of ``s`` (the empty spelling is 0)."""
    rank = 0
    for c in s:
        rank = rank * _N + ord(c)
    return _shorter_count(len(s)) + rank


def spelling_from_index(index: int) -> str:
    """Invert :func:`spelling_index`: the spelling at global shortlex position ``index``."""
    if index <= 0:
        return ""
    length = 1
    while not (_shorter_count(length) <= index < _shorter_count(length + 1)):
        length += 1
    rank = index - _shorter_count(length)
    digits: list[int] = []
    for _ in range(length):
        digits.append(rank % _N)
        rank //= _N
    return "".join(chr(d) for d in reversed(digits))


def _hi_val(hi: int | None) -> float | int:
    """Treat an unbounded upper endpoint (``None``) as positive infinity for comparison."""
    return float("inf") if hi is None else hi


def _subtract_one(
    plo: int, phi: int | None, olo: int, ohi: int | None
) -> list[tuple[int, int | None]]:
    """Return the parts of ``[plo, phi)`` left after removing ``[olo, ohi)``."""
    if _hi_val(ohi) <= plo or olo >= _hi_val(phi):
        return [(plo, phi)]
    out: list[tuple[int, int | None]] = []
    if olo > plo:
        out.append((plo, olo))
    if ohi is not None and _hi_val(phi) > ohi:
        out.append((ohi, phi))
    return out


@dataclass(frozen=True)
class IntervalSet:
    """A normalized set of disjoint half-open shortlex intervals over positions.

    Each interval is ``(lo, hi)`` in global shortlex positions, half-open
    ``[lo, hi)``; ``hi is None`` means unbounded (a final segment). The tuple is
    kept sorted, disjoint, and maximally merged, so equal sets compare equal.
    """

    intervals: tuple[tuple[int, int | None], ...]

    @staticmethod
    def empty() -> IntervalSet:
        """The empty set."""
        return IntervalSet(())

    @staticmethod
    def from_range(lo: str, hi: str | None) -> IntervalSet:
        """The half-open shortlex interval ``[lo, hi)`` (unbounded when ``hi`` is ``None``)."""
        lo_i = spelling_index(lo)
        hi_i = None if hi is None else spelling_index(hi)
        if hi_i is not None and hi_i <= lo_i:
            return IntervalSet.empty()
        return IntervalSet(((lo_i, hi_i),))

    @staticmethod
    def point(s: str) -> IntervalSet:
        """The singleton set ``{s}``."""
        i = spelling_index(s)
        return IntervalSet(((i, i + 1),))

    def is_empty(self) -> bool:
        """Whether the set has no members."""
        return not self.intervals

    def contains(self, s: str) -> bool:
        """Whether ``s`` is a member."""
        i = spelling_index(s)
        return any(lo <= i and (hi is None or i < hi) for lo, hi in self.intervals)

    def cardinality(self) -> int | None:
        """The number of members, or ``None`` when the set is infinite."""
        total = 0
        for lo, hi in self.intervals:
            if hi is None:
                return None
            total += hi - lo
        return total

    def rank(self, s: str) -> int:
        """The number of members strictly less than ``s`` (always finite)."""
        i = spelling_index(s)
        total = 0
        for lo, hi in self.intervals:
            if hi is not None and hi <= i:
                total += hi - lo
            elif lo < i:
                total += i - lo
        return total

    def nth(self, index: int) -> str:
        """The ``index``-th member in spelling order."""
        for lo, hi in self.intervals:
            size = None if hi is None else hi - lo
            if size is None or index < size:
                return spelling_from_index(lo + index)
            index -= size
        msg = "index out of range"
        raise IndexError(msg)

    def union(self, other: IntervalSet) -> IntervalSet:
        """The union of two sets, renormalized."""
        merged: list[tuple[int, int | None]] = []
        for lo, hi in sorted(self.intervals + other.intervals, key=lambda iv: iv[0]):
            if merged and lo <= _hi_val(merged[-1][1]):
                plo, phi = merged[-1]
                new_hi = None if (phi is None or hi is None) else max(phi, hi)
                merged[-1] = (plo, new_hi)
            else:
                merged.append((lo, hi))
        return IntervalSet(tuple(merged))

    def difference(self, other: IntervalSet) -> IntervalSet:
        """The members of this set not in ``other``."""
        result: list[tuple[int, int | None]] = []
        for lo, hi in self.intervals:
            pieces: list[tuple[int, int | None]] = [(lo, hi)]
            for olo, ohi in other.intervals:
                pieces = [part for plo, phi in pieces for part in _subtract_one(plo, phi, olo, ohi)]
            result.extend(pieces)
        return IntervalSet(tuple(result))

    def intersects(self, other: IntervalSet) -> bool:
        """Whether the two sets share any member."""
        return any(
            max(lo, olo) < _hi_val(_min_hi(hi, ohi))
            for lo, hi in self.intervals
            for olo, ohi in other.intervals
        )

    def __iter__(self):  # noqa: ANN204 -- Iterator[str], typed via return of a generator
        """Yield the members in spelling order (lazy; an unbounded set is infinite)."""
        for lo, hi in self.intervals:
            s = spelling_from_index(lo)
            index = lo
            while hi is None or index < hi:
                yield s
                s = successor(s)
                index += 1

    def __repr__(self) -> str:
        """Compact repr: a point interval is its spelling, a range is ``a..z``."""
        if not self.intervals:
            return "{}"
        parts = [_interval_repr(lo, hi) for lo, hi in self.intervals]
        return parts[0] if len(parts) == 1 else "{" + ",".join(parts) + "}"


def _interval_repr(lo: int, hi: int | None) -> str:
    """The compact spelling of a single half-open interval ``[lo, hi)``."""
    lo_s = spelling_from_index(lo)
    if hi is None:
        return f"{lo_s}.."
    if hi - lo == 1:
        return lo_s
    return f"{lo_s}..{spelling_from_index(hi - 1)}"


def _min_hi(a: int | None, b: int | None) -> int | None:
    """The smaller of two upper endpoints, treating ``None`` as infinity."""
    if a is None:
        return b
    if b is None:
        return a
    return min(a, b)

File: rust/src/floor/order.rs
//! The spelling order: shortlex over spellings, and symbolic windows of it.
//!
//! A faithful port of `hejmark/core/floor/order.py`. L1 fixes the spelling
//! order as shortlex (shorter first, ties by code point) -- a well-order of
//! type omega over the whole Unicode code space, surrogates riding along by
//! design, nothing special-cased. This module realizes just what the
//! constructors need:
//!
//! - [`shortlex_cmp`] compares spellings in shortlex.
//! - [`successor`] steps to the unique next spelling, which turns an inclusive
//!   range endpoint into a half-open bound and walks a window lazily.
//! - [`Window`] is a half-open shortlex interval `[lo, hi)`; a range or a final
//!   segment denotes as one window, never materialized. An unbounded window
//!   (`hi` is `None`) still answers `contains` in finite time.
//!
//! One representational note: a Python `str` is a sequence of code points, and
//! the order deliberately ranges over lone surrogates (U+D800..=U+DFFF), which
//! Rust's `char`/`String` cannot hold. So a spelling here is a `[u32]` of code
//! points in `0..=0x10FFFF`, not a `String`. That is the only difference from
//! the Python; the order it induces is identical.

use std::cmp::Ordering;

/// The greatest code point: the one code unit with no in-length successor.
pub const MAX: u32 = 0x10FFFF;

/// Compare two spellings in shortlex: shorter first, ties by code point.
pub fn shortlex_cmp(a: &[u32], b: &[u32]) -> Ordering {
    a.len().cmp(&b.len()).then_with(|| a.cmp(b))
}

/// Return the shortlex successor: the unique next spelling after `s`.
///
/// Increments the last non-maximal code point and zeroes everything after it;
/// an all-maximal spelling rolls over to `s.len() + 1` zeros. The order is
/// total, so every spelling has a successor and the order is walked one step
/// at a time.
pub fn successor(s: &[u32]) -> Vec<u32> {
    let mut codes = s.to_vec();
    for i in (0..codes.len()).rev() {
        if codes[i] != MAX {
            codes[i] += 1;
            for code in codes.iter_mut().skip(i + 1) {
                *code = 0;
            }
            return codes;
        }
    }
    vec![0; s.len() + 1]
}

/// Return whichever spelling is greater in shortlex; ties return `a`.
fn shortlex_max<'a>(a: &'a [u32], b: &'a [u32]) -> &'a [u32] {
    if shortlex_cmp(a, b) == Ordering::Less {
        b
    } else {
        a
    }
}

/// A half-open shortlex window `[lo, hi)`; `hi` is `None` when unbounded.
///
/// A reversed window (`hi` at or below `lo`) is simply empty: [`contains`] finds
/// no member and [`members`] yields nothing.
///
/// [`contains`]: Window::contains
/// [`members`]: Window::members
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Window {
    /// Inclusive lower bound.
    pub lo: Vec<u32>,
    /// Exclusive upper bound, or `None` when the window is unbounded.
    pub hi: Option<Vec<u32>>,
}

impl Window {
    /// Whether the window has no member.
    pub fn is_empty(&self) -> bool {
        match &self.hi {
            Some(hi) => shortlex_cmp(hi, &self.lo) != Ordering::Greater,
            None => false,
        }
    }

    /// Whether `s` falls inside the window.
    pub fn contains(&self, s: &[u32]) -> bool {
        if shortlex_cmp(s, &self.lo) == Ordering::Less {
            return false;
        }
        match &self.hi {
            Some(hi) => shortlex_cmp(s, hi) == Ordering::Less,
            None => true,
        }
    }

    /// The parts of the window left after removing `cut` -- zero, one, or two.
    pub fn minus(&self, cut: &Window) -> Vec<Window> {
        let left_hi = match &self.hi {
            // cut.lo sits at or past hi: nothing to trim on the left, keep hi.
            Some(hi) if shortlex_cmp(&cut.lo, hi) != Ordering::Less => Some(hi.clone()),
            // hi is None, or cut.lo lands inside: the left piece ends at cut.lo.
            _ => Some(cut.lo.clone()),
        };
        let mut pieces = vec![Window {
            lo: self.lo.clone(),
            hi: left_hi,
        }];
        if let Some(cut_hi) = &cut.hi {
            pieces.push(Window {
                lo: shortlex_max(&self.lo, cut_hi).to_vec(),
                hi: self.hi.clone(),
            });
        }
        pieces
            .into_iter()
            .filter(|piece| !piece.is_empty())
            .collect()
    }

    /// Iterate the members in spelling order (lazy; an unbounded window is infinite).
    pub fn members(&self) -> Members {
        Members {
            next: self.lo.clone(),
            hi: self.hi.clone(),
        }
    }
}

/// A lazy iterator over a [`Window`]'s members in spelling order.
///
/// It owns its bound rather than borrowing the window, so it can be moved into
/// and composed by the denotation streams in [`super::universe`] without tying
/// their lifetime to a window that has since gone out of scope.
pub struct Members {
    next: Vec<u32>,
    hi: Option<Vec<u32>>,
}

impl Iterator for Members {
    type Item = Vec<u32>;

    fn next(&mut self) -> Option<Vec<u32>> {
        if let Some(hi) = &self.hi {
            if shortlex_cmp(&self.next, hi) != Ordering::Less {
                return None;
            }
        }
        let current = self.next.clone();
        self.next = successor(&self.next);
        Some(current)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// A spelling as code points, mirroring how the tests write Python `str`s.
    fn cp(s: &str) -> Vec<u32> {
        s.chars().map(|c| c as u32).collect()
    }

    /// Every spelling of length <= 3 over a tiny alphabet, in shortlex order --
    /// small enough to check window operations exhaustively.
    fn universe_of_spellings() -> Vec<Vec<u32>> {
        let alphabet = [b'a' as u32, b'b' as u32, b'c' as u32];
        let mut spellings: Vec<Vec<u32>> = vec![vec![]];
        for &a in &alphabet {
            spellings.push(vec![a]);
        }
        for &a in &alphabet {
            for &b in &alphabet {
                spellings.push(vec![a, b]);
            }
        }
        for &a in &alphabet {
            for &b in &alphabet {
                for &c in &alphabet {
                    spellings.push(vec![a, b, c]);
                }
            }
        }
        spellings.sort_by(|x, y| shortlex_cmp(x, y));
        spellings
    }

    #[test]
    fn shortlex_sorts_shorter_first_then_by_code_point() {
        let mut spellings = vec![cp("ba"), cp("b"), cp("a"), cp(""), cp("ab")];
        spellings.sort_by(|x, y| shortlex_cmp(x, y));
        assert_eq!(
            spellings,
            vec![cp(""), cp("a"), cp("b"), cp("ab"), cp("ba")]
        );
    }

    #[test]
    fn successor_is_strictly_next() {
        for s in universe_of_spellings() {
            assert_eq!(shortlex_cmp(&s, &successor(&s)), Ordering::Less);
        }
        // The rollover cases the tiny universe never reaches.
        for s in [vec![], vec![MAX], vec![MAX, MAX]] {
            assert_eq!(shortlex_cmp(&s, &successor(&s)), Ordering::Less);
        }
    }

    #[test]
    fn successor_steps_within_a_length_then_rolls_over() {
        assert_eq!(successor(&cp("")), cp("\u{0}"));
        assert_eq!(successor(&cp("a")), cp("b"));
        assert_eq!(successor(&cp("az")), cp("a{"));
        assert_eq!(successor(&cp("ab\u{10ffff}")), cp("ac\u{0}"));
    }

    #[test]
    fn all_maximal_spelling_rolls_to_a_longer_zero_spelling() {
        assert_eq!(successor(&[MAX]), vec![0, 0]);
        assert_eq!(successor(&[MAX, MAX]), vec![0, 0, 0]);
    }

    #[test]
    fn bounded_window_iterates_in_spelling_order() {
        let window = Window {
            lo: cp("a"),
            hi: Some(successor(&cp("e"))),
        };
        let members: Vec<Vec<u32>> = window.members().collect();
        assert_eq!(members, vec![cp("a"), cp("b"), cp("c"), cp("d"), cp("e")]);
    }

    #[test]
    fn window_contains_matches_brute_force() {
        let window = Window {
            lo: cp("b"),
            hi: Some(cp("bb")),
        };
        for s in universe_of_spellings() {
            let expected = shortlex_cmp(&cp("b"), &s) != Ordering::Greater
                && shortlex_cmp(&s, &cp("bb")) == Ordering::Less;
            assert_eq!(window.contains(&s), expected, "spelling {s:?}");
        }
    }

    #[test]
    fn unbounded_window_is_infinite() {
        let tail = Window {
            lo: cp("a"),
            hi: None,
        };
        assert!(tail.contains(&cp("a")));
        assert!(tail.contains(&cp("zzzz")));
        assert!(!tail.contains(&cp("\u{0}"))); // below "a" in shortlex
        let first_three: Vec<Vec<u32>> = tail.members().take(3).collect();
        assert_eq!(first_three, vec![cp("a"), cp("b"), cp("c")]);
    }

    #[test]
    fn reversed_window_is_empty() {
        let window = Window {
            lo: cp("e"),
            hi: Some(cp("a")),
        };
        assert!(window.is_empty());
        assert!(!window.contains(&cp("e")));
        assert!(window.members().next().is_none());
    }

    #[test]
    fn minus_matches_brute_force() {
        let universe = universe_of_spellings();
        let windows = [
            Window {
                lo: cp("a"),
                hi: Some(cp("cc")),
            },
            Window {
                lo: cp("b"),
                hi: None,
            },
        ];
        let cuts = [
            Window {
                lo: cp("b"),
                hi: Some(cp("bc")),
            },
            Window {
                lo: cp("aa"),
                hi: Some(cp("ca")),
            },
            Window {
                lo: cp("e"),
                hi: Some(cp("a")), // an empty cut removes nothing
            },
        ];
        for window in &windows {
            for cut in &cuts {
                let pieces = window.minus(cut);
                for s in &universe {
                    let in_pieces = pieces.iter().any(|piece| piece.contains(s));
                    let expected = window.contains(s) && !cut.contains(s);
                    assert_eq!(
                        in_pieces, expected,
                        "window {window:?} minus {cut:?} at {s:?}"
                    );
                }
            }
        }
    }
}
